storage/aptosdb/src/state_store/mod.rs

```diff
@@ -63,6 +72,88 @@ impl StateStore {
         ))
     }
 
+    #[cfg(test)]
+    fn get_node_keys_by_key_prefix(
+        &self,
+        key_prefix: &StateKeyPrefix,
+        desired_version: Version,
+    ) -> Result<HashMap<StateKey, NodeKey>> {
+        let mut iter = self.db.iter::<StateValueIndexSchema>(Default::default())?;
+        let mut result = HashMap::new();
+        iter.seek(&(key_prefix))?;
+        while let Some(((state_key, first_version), num_nibbles)) = iter.next().transpose()? {
+            // Cursor is currently at the first available version of the state key.
+            // Check if the key_prefix is a valid prefix of the state_key we got from DB.
+
+            if !key_prefix.is_prefix(&state_key)? {
+                // No more keys matching the key_prefix, we can return the result.
+                return Ok(result);
+            }
+            match first_version.cmp(&desired_version) {
+                Ordering::Less => {
+                    iter.seek_for_prev(&(state_key.clone(), desired_version))?;
+                    let ((state_key, db_version), num_nibbles) =
+                        iter.next().transpose()?.ok_or_else(|| {
+                            anyhow!(
+                                "Failure seeking to desired version {:?} for state key {:?}",
+                                desired_version,
+                                state_key
+                            )
+                        })?;
+                    result.insert(
+                        state_key.clone(),
+                        NodeKey::new(
+                            db_version,
+                            NibblePath::new_from_state_key(&state_key, num_nibbles as usize),
+                        ),
+                    );
+                }
+
+                Ordering::Equal => {
+                    result.insert(
+                        state_key.clone(),
+                        NodeKey::new(
+                            first_version,
+                            NibblePath::new_from_state_key(&state_key, num_nibbles as usize),
+                        ),
+                    );
+                }
+                Ordering::Greater => {}
+            }
```

Comment:
will be a few lines shorter if we do
```
if first_version <= desired_version {
    if first_version < desired_version {
        seek(..)
    }
    insert(..)
}
```

Reply:
That's true but my preference would be to avoid multi-level `if` conditions if possible, as that hurts the readability of the code. Would prefer to keep it this way unless you have a strong opinion.